node/src/tasks/bsp_upload_file_task.rs

```diff
@@ -0,0 +1,339 @@
+                event.chunk_with_proof.proven.key, event.file_key
+            )));
+        }
+
+        let mut write_file_storage = self.storage_hub_handler.file_storage.write().await;
+        let write_chunk_result = write_file_storage.write_chunk(
+            &file_key,
+            &event.chunk_with_proof.proven.key,
+            &event.chunk_with_proof.proven.data,
+        );
+        // Release the file storage write lock as soon as possible.
+        drop(write_file_storage);
+
+        match write_chunk_result {
+            Ok(outcome) => match outcome {
+                FileStorageWriteOutcome::FileComplete => self.on_file_complete(&file_key).await,
+                FileStorageWriteOutcome::FileIncomplete => {}
+            },
+            Err(error) => match error {
+                FileStorageWriteError::FileChunkAlreadyExists => {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Received duplicate chunk with key: {}",
+                        event.chunk_with_proof.proven.key
+                    );
+
+                    // TODO: Consider informing this to the file transfer service so that it can handle reputation for this peer id.
+                }
+                FileStorageWriteError::FileDoesNotExist => {
+                    // Unvolunteer the file.
+                    self.unvolunteer_file(file_key).await?;
+
+                    return Err(anyhow::anyhow!(format!("File does not exist for key {:?}. Maybe we forgot to unregister before deleting?", event.file_key)));
+                }
+                FileStorageWriteError::FailedToGetFileChunk
+                | FileStorageWriteError::FailedToInsertFileChunk => {
+                    // This internal error should not happen.
+
+                    // Unvolunteer the file.
+                    self.unvolunteer_file(file_key).await?;
+
+                    return Err(anyhow::anyhow!(format!(
+                        "Internal trie read/write error {:?}:{}",
+                        event.file_key, event.chunk_with_proof.proven.key
+                    )));
+                }
+                FileStorageWriteError::FingerprintAndStoredFileMismatch => {
+                    // This should never happen, given that the first check in the handler is verifying the proof.
+                    // This means that something is seriously wrong, so we error out the whole task.
+
+                    // Unvolunteer the file.
+                    self.unvolunteer_file(file_key).await?;
+
+                    return Err(anyhow::anyhow!(format!(
+                        "Invariant broken! This is a bug! Fingerprint and stored file mismatch for key {:?}.",
+                        event.file_key
+                    )));
+                }
+            },
+        }
+
+        Ok(())
+    }
+}
+
+impl<SHC: StorageHubHandlerConfig> EventHandler<NewStorageRequest> for BspUploadFileTask<SHC> {
```

Comment:
I would move this more to the top of this file, as this is the first event that is handled by the flow of this task.

Reply:
Also, it could be worth to comment what part of the flow of this task this event is handling.